### PR TITLE
museum away mission

### PR DESCRIPTION
**Pull request by mc-oofert**
## About The Pull Request

adds a new gateway map, the Nanotrasen Museum it is filled with """"Mannequins"""" and Common Core lore
im not putting the preview here because you really should explore it yourself but if youre that curious i think the Checks tab in mapdiffbot would have it
this gateway map contains no combat unless you count falling into chasms because you did not carry a light
or going into the boarded room with no loot or any incentive with obvious signs that there is the sole enemy on the map in there
the loot is the lore ok thanks

also makes mines detonate if theyre detonated by a non-mob im pretty sure this couldnt have been intentional
trams stop chasms
and also the relevant items

<details>
  <summary>on second thought if you want spoilers check this</summary>

image

</details>

## Why It's Good For The Game

more gateway maps = good

## Changelog
:cl:
add: nanotrasen museum gateway map
/:cl:

### CODE REVIEW

File: code/modules/mapfluff/ruins/objects_and_mobs/museum.dm [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	var/target_turf_x = 0
+	var/target_turf_y = 0
+	/// if not specified, uses our Z
+	var/target_turf_z
+
+/turf/open/mirage/Initialize(mapload)
+	. = ..()
+	if(isnull(range))
+		range = world.view
+	var/used_z = target_turf_z ? target_turf_z : z //if target z is not defined, use ours
```
Comment (MrMelbert) [inline code/modules/mapfluff/ruins/objects_and_mobs/museum.dm:84]:
```suggestion
	var/used_z = target_turf_z || z //if target z is not defined, use ours
```

File: code/modules/mapping/mapping_helpers.dm [hunk 2]
```diff
@@ -1395,3 +1395,33 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
...
+	var/atom/movable/buckle_to
+	var/list/mobs = list()
+	for(var/thing as anything in loc)
+		var/atom/movable/possible_buckle = thing
+		if(isnull(buckle_to) && istype(possible_buckle) && possible_buckle.can_buckle)
+			buckle_to = possible_buckle
+			continue
+
+		if(isliving(thing))
+			mobs += thing
```
Comment (MrMelbert) [inline code/modules/mapping/mapping_helpers.dm:1418]:
```suggestion
	for(var/atom/movable/possible_buckle as anything in loc)
		if(isnull(buckle_to) && possible_buckle.can_buckle)
			buckle_to = possible_buckle
			continue

		if(isliving(possible_buckle))
			mobs += possible_buckle 
```

File: code/modules/mapping/mapping_helpers.dm [hunk 3]
```diff
@@ -1395,3 +1395,33 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
...
+	
+	if(isnull(buckle_to))
+		log_mapping("[type] at [x] [y] [z] did not find anything to buckle to")
+		return INITIALIZE_HINT_QDEL
+		
+	for(var/mob/living/mob in mobs)
+		buckle_to.buckle_mob(mob, force = force_buckle)
```
Comment (MrMelbert) [inline code/modules/mapping/mapping_helpers.dm:1425]:
```suggestion
	for(var/mob/living/mob as anything in mobs)
		buckle_to.buckle_mob(mob, force = force_buckle)
```

File: code/__DEFINES/traits/declarations.dm [hunk 4]
```diff
@@ -602,6 +602,8 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_IMMERSE_STOPPED "immerse_stopped"
 /// The effects of hyperspace drift are blocked when the tile has this trait
 #define TRAIT_HYPERSPACE_STOPPED "hyperspace_stopped"
+///Chasms will be safe to cross if there is something with this trait on it
+#define TRAIT_CHASM_STOPPER "chasm_stopper"
```
Comment (MrMelbert) [inline code/__DEFINES/traits/declarations.dm:606]: Put this one two defines up right by `TRAIT_CHASM_STOPPED`

Comment (MrMelbert) [reply code/__DEFINES/traits/declarations.dm:606]: Also, with the existence of [`TRAIT_CHASM_STOPPED`], you may not even need [`TRAIT_CHASM_STOPPER`]? 

You can just give atoms `TRAIT_CHASM_STOPPED` when they enter a tram and remove it when they exit, I think you can do it in `add_item_on_transport` / `remove_item_from_transport`

File: code/game/objects/items/puzzle_pieces.dm [hunk 5]
```diff
@@ -403,3 +403,61 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/puzzle_keycardpad, 32)
...
+	light_range = 1
+	density = FALSE
+
+/obj/structure/puzzle_blockade/oneway/update_icon_state()
+	icon_state = "[base_icon_state][density ? "" : "-off"]"
+	return ..()
+
+/obj/structure/puzzle_blockade/oneway/CanAllowThrough(atom/movable/mover, border_dir)
+	. = ..()
+	return . && (REVERSE_DIR(border_dir) == dir || get_turf(mover) == get_turf(src))
```
Comment (MrMelbert) [inline code/game/objects/items/puzzle_pieces.dm:423]:
```suggestion
	return ..() && (REVERSE_DIR(border_dir) == dir || get_turf(mover) == get_turf(src))
```

File: code/modules/mapfluff/ruins/objects_and_mobs/museum.dm [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	density = TRUE
+	opacity = FALSE
+	var/replica_path = /obj/structure/fluff
+	var/target_path
+	var/obvious_replica = TRUE
+
+/obj/effect/replica_spawner/Initialize(mapload)
+	. = ..()
+	INVOKE_ASYNC(src, PROC_REF(create_replica))
+	return INITIALIZE_HINT_QDEL
```
Comment (MrMelbert) [inline code/modules/mapfluff/ruins/objects_and_mobs/museum.dm:25]: Why async?

File: code/modules/mob/living/basic/space_fauna/statue/mannequin.dm [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+/mob/living/basic/statue/mannequin/Initialize(mapload)
+	. = ..()
+	update_appearance()
+
+/mob/living/basic/statue/mannequin/update_overlays()
+	. = ..()
+	if(held_item)
+		. += mutable_appearance(initial(held_item.righthand_file), initial(held_item.inhand_icon_state))
+	if(hat)
+		. += mutable_appearance(initial(hat.worn_icon), initial(hat.worn_icon_state) ? initial(hat.worn_icon_state) : initial(hat.icon_state))
```
Comment (MrMelbert) [inline code/modules/mob/living/basic/space_fauna/statue/mannequin.dm:28]:
```suggestion
		. += mutable_appearance(initial(hat.worn_icon), initial(hat.worn_icon_state) || initial(hat.icon_state))
```

File: code/game/objects/items/puzzle_pieces.dm [hunk 8]
```diff
@@ -403,3 +403,61 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/puzzle_keycardpad, 32)
...
+		RegisterSignal(partner, COMSIG_PUZZLE_COMPLETED, PROC_REF(try_signal))
+
+/obj/effect/puzzle_poddoor_open/proc/try_signal(datum/source)
+	SIGNAL_HANDLER
+	var/openclose
+	for(var/obj/machinery/door/poddoor/door as anything in SSmachines.get_machines_by_type_and_subtypes(/obj/machinery/door/poddoor))
+		if(door.id == id)
+			if(openclose == null)
+				openclose = door.density
+			INVOKE_ASYNC(door, openclose ? TYPE_PROC_REF(/obj/machinery/door/poddoor, open) : TYPE_PROC_REF(/obj/machinery/door/poddoor, close))
```
Comment (MrMelbert) [inline code/game/objects/items/puzzle_pieces.dm:463]:
```suggestion
		if(door.id != id)
			continue
		if(isnull(openclose))
			openclose = door.density
		INVOKE_ASYNC(door, openclose ? TYPE_PROC_REF(/obj/machinery/door/poddoor, open) : TYPE_PROC_REF(/obj/machinery/door/poddoor, close))
```

File: code/__DEFINES/traits/declarations.dm (hunk 4 above)
Comment (mc-oofert) [reply code/__DEFINES/traits/declarations.dm:606]: chasm_stopped is for turfs

**Comment by Loafin34** [comment]
two things
1. cortical borer 
2. gyro
good job mc-oofert never change

**Comment by Wallemations** [comment]
I'll admit I'm not entirely familiar with how gateway areas work; does the map need to be so big if only a certain portion has actual station stuff in it?

**Comment by Xkallubar** [comment]
Time for some rambling!

This seems cool, but I'd prefer it be bigger, have areas with actual enemies, and have worthwhile rewards.

The moon outpost gateway has live facehugger eggs
The research gateway has an armory full of guns and armor
The snowcabin gateway has a fully functional nukie modsuit
The other gateways that are still enabled don't have any worthwhile rewards imo.

This just has some lore and funny cortical borer in a box.

Maybe put down something useful, like a construct shell in a cult exhibit for any chaplains that happened to have a shade, or a dead morph that could be revived and made sentient for science shenanigans.

Put some loose possibilities of things you could do with what's in this gateway. 

Pieces to a puzzle, like as mentioned above, a construct shell, or the corpse of a deprecated antag.

What happens when the beach gateway shows up? Not much. Assistants flood in to just grab the free medkits and leave.

What happens when the moon outpost gateway shows up? Science almost always does an expedition to catch a facehugger.

What will happen when everyone's been to your museum gateway?
Will people even bother?
Is there good enough rewards to warrant entering?

**Comment by mc-oofert** [comment]
most of what you said goes against the rules of the bounty

**Comment by Admiral-Gigaton** [comment]
sorry im not too familiar with mapping, but

whats the pod in the top left corner of the structure?

and where is the "funny man" in a box, i cant seem to find him anywhere, though i might just be brain damaged

**Comment by mc-oofert** [comment]
> whats the pod in the top left corner of the structure?

structure is a wide term you have to be more precise

> and where is the "funny man" in a box, i cant seem to find him anywhere, though i might just be brain damaged

what box

> otherwise, i like this, i like this a lot, good gimmicks could happen, though people could just cheese it with a stack of rods by covering the chasms, sooo... also, adding duct spiders isnt a bad idea for flavor, this IS a ruined museum after all, some little critters are bound to make their home there, right?

i would say that most living things there """"mysteriously"""" """"""disappeared"""""", it isnt ruined just 98% devoid of life

**Comment by Singul0** [comment]
> Time for some rambling!
> 
> This seems cool, but I'd prefer it be bigger, have areas with actual enemies, and have worthwhile rewards.
> 
> The moon outpost gateway has live facehugger eggs The research gateway has an armory full of guns and armor The snowcabin gateway has a fully functional nukie modsuit The other gateways that are still enabled don't have any worthwhile rewards imo.
> 
> This just has some lore and funny cortical borer in a box.
> 
> Maybe put down something useful, like a construct shell in a cult exhibit for any chaplains that happened to have a shade, or a dead morph that could be revived and made sentient for science shenanigans.
> 
> Put some loose possibilities of things you could do with what's in this gateway.
> 
> Pieces to a puzzle, like as mentioned above, a construct shell, or the corpse of a deprecated antag.
> 
> What happens when the beach gateway shows up? Not much. Assistants flood in to just grab the free medkits and leave.
> 
> What happens when the moon outpost gateway shows up? Science almost always does an expedition to catch a facehugger.
> 
> What will happen when everyone's been to your museum gateway? Will people even bother? Is there good enough rewards to warrant entering?

eh, it's pretty good for what the bounty asks for in gateway maps. you can't really add much without a maintainer asking you to rebalance it these days anyways. You need to understand that the maintainers take balancing really seriously spreadsheets and all. I love all the little effects and gimmicks the map has in it. nice, compactt and ripe for expansions.

File: code/__DEFINES/traits/declarations.dm (hunk 4 above)
Comment (Ghommie) [reply code/__DEFINES/traits/declarations.dm]: Can't you use the `give_turf_traits` element instead of adding a new redundant trait?

Comment (mc-oofert) [reply code/__DEFINES/traits/declarations.dm]: Trust me I wouldnt do it if it worked

Comment (Ghommie) [reply code/__DEFINES/traits/declarations.dm]: I see. Tram movement is a bit niche.

**Comment by MMMiracles** [comment]
dark spots:
image
image
image
image
image
I'm not sure how intentional these are, considering your use of dimmed lights. They seem too dark in-game though.

After running around the map locally, I personally think this is a pretty neat and succinct walk-through of both actual lore and some actual game changes. The terminal descriptions for all the exhibits seem decent enough, keeps any balance/content changes fairly in-world when explaining them so that's a plus. The mirage hallways and seamless teleports to wrap around is very unique and honestly hard to see the transition unless you're really looking for it. Nice touch of flair and its another addition to future potential maps. The loot itself certainly is on the lower side but with how inherently non-lethal the mission it, I say it fits a nice niche in the peaceful category ala beach but with something new to look at.

Actual mapping wise, since this is a much more contained experience, I would recommend maybe adding some decal flair to help further mimic a proper outpost/station area. Fluff floor vents/scrubbers, fake wall panels/cameras, maybe some announcement screens? It would help a lot with how barren the walls feel in some spots and help better sell it as an actual location you'd play in. All the exhibits are fairly succinct and get their point across well along with the terminal context. As I said earlier, the mirage hallways are pretty well done and honestly had me confused for a second when I had strongdmm and the local pulled up together trying to figure out where the hell you mapped the hallway. The breakable walls are a neat concept but I'd look into making the fuel puddles a lot more noticeable, I didn't even see them until I right-clicked the false wall and saw the puddle on the list. 

Balance wise, I didn't see anything crazy in there. Some language books, bit of maintenance loot scattered about and a (decent?) stack of space cash. The secret is pretty fun and keeps the mission mostly peaceful unless you intentionally let it out. I would recommend upgrading those catwalks to something that can't be cut. As soon as I found the wirecutters my first thought was "i can just completely ruin the secret bridge with like 3 snips". Sure it can be fixed but that just seems unnecessary but getting dunked into an away mission chasm (un)intentionally just sounds like not a great time. 

Seems pretty good to me, if everything is good code-wise I don't see any major issues with the map itself or how it generally runs. I would mainly just like to see some additional station fluff ala cameras/wall mounts to help cover up the walls and make the hall areas seem more busy.

**Comment by mc-oofert** [comment]
> -snip-
> 
> After running around the map locally, I personally think this is a pretty neat and succinct walk-through of both actual lore and some actual game changes. The terminal descriptions for all the exhibits seem decent enough, keeps any balance/content changes fairly in-world when explaining them so that's a plus. The mirage hallways and seamless teleports to wrap around is very unique and honestly hard to see the transition unless you're really looking for it. Nice touch of flair and its another addition to future potential maps. The loot itself certainly is on the lower side but with how inherently non-lethal the mission it, I say it fits a nice niche in the peaceful category ala beach but with something new to look at.
> 
> Actual mapping wise, since this is a much more contained experience, I would recommend maybe adding some decal flair to help further mimic a proper outpost/station area. Fluff floor vents/scrubbers, fake wall panels/cameras, maybe some announcement screens? It would help a lot with how barren the walls feel in some spots and help better sell it as an actual location you'd play in. All the exhibits are fairly succinct and get their point across well along with the terminal context. As I said earlier, the mirage hallways are pretty well done and honestly had me confused for a second when I had strongdmm and the local pulled up together trying to figure out where the hell you mapped the hallway. The breakable walls are a neat concept but I'd look into making the fuel puddles a lot more noticeable, I didn't even see them until I right-clicked the false wall and saw the puddle on the list.
> 
> Balance wise, I didn't see anything crazy in there. Some language books, bit of maintenance loot scattered about and a (decent?) stack of space cash. The secret is pretty fun and keeps the mission mostly peaceful unless you intentionally let it out. I would recommend upgrading those catwalks to something that can't be cut. As soon as I found the wirecutters my first thought was "i can just completely ruin the secret bridge with like 3 snips". Sure it can be fixed but that just seems unnecessary but getting dunked into an away mission chasm (un)intentionally just sounds like not a great time.
> 
> Seems pretty good to me, if everything is good code-wise I don't see any major issues with the map itself or how it generally runs. I would mainly just like to see some additional station fluff ala cameras/wall mounts to help cover up the walls and make the hall areas seem more busy.

done

File: code/modules/mob/living/basic/vermin/mothroach.dm [hunk 9]
```diff
@@ -73,3 +73,10 @@
 	planning_subtrees = list(
 		/datum/ai_planning_subtree/random_speech/mothroach,
 	)
+
+/mob/living/basic/mothroach/bar
+	name = "mothroach bartender"
+	desc = "Mothroach serving drinks look at him go"
```
Comment (Ghommie) [inline code/modules/mob/living/basic/vermin/mothroach.dm:79]:
```suggestion
	desc = "A mothroach serving drinks. Look at him go."
```

File: code/modules/mapfluff/ruins/objects_and_mobs/museum.dm [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+/turf/open/mirage/Initialize(mapload)
+	. = ..()
+	if(isnull(range))
+		range = world.view
+	var/used_z = target_turf_z || z //if target z is not defined, use ours
+	var/turf/target = locate(offset ? target_turf_x + x : target_turf_x, offset ? target_turf_y + y : target_turf_y, used_z)
+	AddElement(/datum/element/mirage_border, target, dir, range)
+
+/obj/effect/mapping_helpers/ztrait_injector/museum
+	traits_to_add = list(ZTRAIT_NOPARALLAX = TRUE,ZTRAIT_NOXRAY = TRUE, ZTRAIT_NOPHASE = TRUE, ZTRAIT_BASETURF = /turf/open/indestructible/plating, ZTRAIT_SECRET = TRUE)
```
Comment (Ghommie) [inline code/modules/mapfluff/ruins/objects_and_mobs/museum.dm:110]:
```suggestion
	traits_to_add = list(ZTRAIT_NOPARALLAX = TRUE, ZTRAIT_NOXRAY = TRUE, ZTRAIT_NOPHASE = TRUE, ZTRAIT_BASETURF = /turf/open/indestructible/plating, ZTRAIT_SECRET = TRUE)
```

File: code/modules/mapfluff/ruins/objects_and_mobs/museum.dm [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+	name = "exhibit info terminal"
+	desc = "A relatively low-tech info board. Not as low-tech as an actual sign though. Appears to be quite old."
+	upperinfo = "Nanotrasen Museum Exhibit Info"
+	icon_state = "plaque"
+	icon_screen = "plaque_screen"
+	icon_keyboard = null
+
+/obj/effect/replica_spawner //description and name are intact, better to make a new fluff object for stuff that is not actually ingame as an object
+	name = "replica creator"
+	desc = "This creates a fluff object that looks exactly like the input, but like obviously a replica. Do not for the love of god use with stuff that has Initialize sideeffects."
```
Comment (Ghommie) [inline code/modules/mapfluff/ruins/objects_and_mobs/museum.dm:11]:
```suggestion
	desc = "This creates a fluff object that looks exactly like the input, but like obviously a replica. Do not for the love of god use with stuff that has Initialize side effects."
```

**Comment by Ghommie** [comment]
Do you guys think this is ready to go?

**Comment by MMMiracles** [comment]
> Do you guys think this is ready to go?

My issues with the map were resolved, if the code is good then I approve.

**Pull request merged**